Use the built-in JSON library

elfenlaid opened the pull request:
Address https://github.com/phoenixframework/phoenix/issues/6461

The changes set the built-in [JSON library](https://hexdocs.pm/elixir/JSON.html) as a default JSON library.

The idea is to check whether the built-in JSON library is present and use it if so:

```elixir
  @default_json_library if Code.ensure_loaded?(JSON), do: JSON, else: Jason

  def json_library do
    Application.get_env(:phoenix, :json_library, @default_json_library)
  end
```

Correct me if I'm wrong, but I think a naive check of whether the library is present during compilation time should work, as it's a built-in library.

---

What I'm not sure about is whether such a move results in a breaking change. 

It could be a breaking change for projects that don't specify the `json_library` and use `@derive Jason.Encoder` on structures passed to `Phoenix.Controller.json/2`. That sounds pretty niche, especially given that the template project includes:

```elixir
config :phoenix, :json_library, Jason
```

Also, I'm specifically leaving the `jason` dependency in place for pre-27 OTP versions.

---

Please let me know what you think and whether the changes make sense. In other words, maybe it would be better to stick with `Jason` until Phoenix requires OTP 28.

---

Similar discussions / changes:

[Use Jason instead of Poison for json encoding by chrismccord #2734phoenixframework/phoenix](https://github.com/phoenixframework/phoenix/pull/2734)
[Investigate using Jason instead of Poison #2693phoenixframework/phoenix](https://github.com/phoenixframework/phoenix/issues/2693)

[Use JSON instead of Jason by kubosuke #234felt/geo](https://github.com/felt/geo/pull/234)

elfenlaid committed:
Use the built-in JSON library by default

elfenlaid committed:
Update documentation

elfenlaid commented on guides/controllers.md:
```diff
@@ -94,7 +94,7 @@ end
 Now [`/hello/Frank`] in your browser should display `From messenger Frank` as plain text without any HTML.
 
-A step beyond this is rendering pure JSON with the [`json/2`] function. We need to pass it something that the [Jason library](`Jason`) can decode into JSON, such as a map. (Jason is one of Phoenix's dependencies.)
+A step beyond this is rendering pure JSON with the [`json/2`] function. We need to pass it something that the [JSON library](https://hexdocs.pm/elixir/JSON.html) can decode into JSON, such as a map.
```
Here's how it looks like with mouse hover:

![Image](https://github.com/user-attachments/assets/ca01f491-281b-42a8-bc9c-e53d4f1c7e96)

elfenlaid commented on config/config.exs:
```diff
@@ -5,7 +5,7 @@ config :logger, :console,
   format: "\n$time $metadata[$level] $message\n"
 
 config :phoenix,
-  json_library: Jason,
```
Probably here we would need to have something similar to `if Code.ensure_loaded?(JSON), do: JSON, else: Jason` if that conditional loading trick makes sense.

elfenlaid committed:
Keep Jason as the default runtime library

elfenlaid committed:
Determine the dev json library at runtime

elfenlaid committed:
Adopt tests to support both JSON and Jason libraries

josevalim reviewed (approved):
Yes, v1.9 material for sure. But it looks good. We can probably assign a v1.9 milestone and merge it once we finally branch v1.8 out. I don’t think it is worth branching out though, as I dont think there is anything beyond this PR for now?

elfenlaid committed:
Ignore json assertion messages

SMC17 commented:
Approval already landed, so this is just a small confirmation pass from a Phoenix + Bandit + LiveView consumer.

The runtime `Code.ensure_loaded?(JSON)` switch in `config/config.exs` reads correctly against the CI matrix: `1.15.8 / OTP 25.3.2.9` resolves to `Jason`, `1.18.4 / OTP 27.3` and `1.18.4 / OTP 28.0.1` resolve to `JSON`. Both branches of the conditional are exercised on every CI run because the matrix straddles the cutoff, which is exactly the case `warn_on_missing_json_library` needs to prove out and does.

The `Phoenix.json_library/0` default in `lib/phoenix.ex` stays on `Jason` with a `v2.0` TODO comment. For library callers that depend on `Phoenix.json_library/0` without a config override, the behavior is unchanged on 1.9, which is the right backward-compat choice. Worth noting that new projects generated by the 1.9 installer get the explicit `config :phoenix, :json_library, JSON` line, so the default codepath in `Phoenix.json_library/0` is only reached by existing apps or by users who delete the line. Both are fine; just recording the surface for posterity.

The test refactor in `test/phoenix/test/conn_test.exs` to derive `json_error` from `Phoenix.json_library()` is cleaner than the previous hardcoded `Jason.DecodeError` assertion; it generalises to any future configured decoder that raises a typed exception. Same for the v1/v2 serializer tests routing through `Phoenix.json_library().decode!/1`.

@SteffenDE's point on the installer Elixir requirement is the one open thread: `installer/mix.exs` still pins `@elixir_requirement "~> 1.15"` and the comment block above it enumerates the five files that move together. Reading the existing tag policy and the "v1.9 milestone, merge after v1.8 branch-out" framing from @josevalim, my read is the requirement bump belongs in a separate 1.9-milestone PR rather than being folded in here, but I can see the argument either way. Is the intent to ship #6481 as-is and open a paired PR for the installer floor bump, or to extend this PR before v1.8 branches out?

Either way, this is a clean change. Thanks for putting it together @elfenlaid, and thanks @josevalim for the v1.9 milestone signal.

saleyn commented:
What would be a downside of providing a JSON behavior for phoenix?

josevalim commented:
The issue is that Ecto has its own library, Ash as well, and so on. If each of them need a behavior, you would then need all of them to agree or you end up with several wrappers. And if they are going to agree on one, the Elixir interface will most likely be the one to choose.

rhcarvalho commented on guides/controllers.md:
```diff
@@ -94,7 +94,7 @@ end
 Now [`/hello/Frank`] in your browser should display `From messenger Frank` as plain text without any HTML.
 
-A step beyond this is rendering pure JSON with the [`json/2`] function. We need to pass it something that the [Jason library](`Jason`) can decode into JSON, such as a map. (Jason is one of Phoenix's dependencies.)
+A step beyond this is rendering pure JSON with the [`json/2`] function. We need to pass it something that the [JSON library](https://hexdocs.pm/elixir/JSON.html) can decode into JSON, such as a map.
```
```suggestion
A step beyond this is rendering pure JSON with the [`json/2`] function. We need to pass it something that the [JSON library](https://elixir.hexdocs.pm/JSON.html) can decode into JSON, such as a map.
```

elfenlaid replied:
Thank you! ❤️

rhcarvalho commented on guides/json_and_apis.md:
````diff
@@ -171,7 +171,7 @@ defmodule HelloWeb.UrlJSON do
 ```
 
-This view is very simple. The `index` function receives all URLs, and converts them into a list of maps. Those maps are placed inside the data key at the root, exactly as we saw when interfacing with our application from `cURL`. In other words, our JSON view converts our complex data into simple Elixir data-structures. Once our view layer returns, Phoenix uses the `Jason` library to encode JSON a ...
+This view is very simple. The `index` function receives all URLs, and converts them into a list of maps. Those maps are placed inside the data key at the root, exactly as we saw when interfacing with our application from `cURL`. In other words, our JSON view converts our complex data into simple Elixir data-structures. Once our view layer returns, Phoenix uses the built-in [JSON library](https:// ...
````
```suggestion
This view is very simple. The `index` function receives all URLs, and converts them into a list of maps. Those maps are placed inside the data key at the root, exactly as we saw when interfacing with our application from `cURL`. In other words, our JSON view converts our complex data into simple Elixir data-structures. Once our view layer returns, Phoenix uses the built-in [JSON library](https://elixir.hexdocs.pm/JSON.html) to encode JSON and send the response to the client.
```

elfenlaid replied:
Thank you! ❤️

rhcarvalho commented on installer/templates/phx_single/config/config.exs.eex:
```diff
@@ -59,8 +59,8 @@ config :logger, :default_formatter,
-# Use Jason for JSON parsing in Phoenix
-config :phoenix, :json_library, Jason
+# Use built-in `JSON` for JSON parsing in Phoenix
+config :phoenix, :json_library, JSON
```
I think here we could do a similar runtime detection as in https://github.com/phoenixframework/phoenix/pull/6481/changes#r2356177940, and only configure `JSON` as the default for Elixir versions where its available.

Same for the umbrella app config template.

elfenlaid replied:
It's a nice idea, yet I don't think it would fly.

The first option is to check Elixir version during `phx.new`, but the Elixir version that generates the project is not guaranteed to match the Elixir version that runs the project. So it sounds like a breaking change.

The second option is to embed this check into new projects. Technically it should work, but I'm not sure that such checks are awkward and probably are not something you would like to see in new projects.

SteffenDE replied:
I think we'll need to bump the elixir requirement for the generated app to 1.18.

saleyn commented on lib/phoenix.ex:
```diff
@@ -45,7 +47,7 @@ defmodule Phoenix do
   """
   def json_library do
-    Application.get_env(:phoenix, :json_library, Jason)
+    Application.get_env(:phoenix, :json_library, @default_json_library)
```
Consider :persistent_term, which is roughly 30-100% faster:

```elixir
defmodule JSON.Adapter do
  @default_json_library if Code.ensure_loaded?(JSON), do: JSON, else: Jason

  def json_library do
    case :persistent_term.get(:json_library, nil) do
      nil ->
        module = :application.get_env(:phoenix, :json_library, @default_json_library)
        :persistent_term.put(:json_library, module)
        module
      module ->
        module
    end
  end
end
```

elfenlaid replied:
TIL `persistent_term` and it looks cool, but I'm not sure if we are in the position of adding anything significant to this PR. I'm happy to add it, but it's @SteffenDE call at this point

rhcarvalho reviewed (commented):
In the meantime HexDocs URLs now use subdomains.

The installer could use runtime detection to decide on the JSON library for new apps, that will keep the installer compatible with Elixir pre-1.18.

elfenlaid committed:
Update guides/json_and_apis.md

Co-authored-by: Rodolfo Carvalho <rhcarvalho@gmail.com>

elfenlaid committed:
Update guides/controllers.md

Co-authored-by: Rodolfo Carvalho <rhcarvalho@gmail.com>

SteffenDE committed:
merge main

rhcarvalho commented:
_I'm casually going through older open PRs to help unblock/triage some low-hanging fruit._

> We should also remove `{:jason, "~> 1.2"}` from new apps.

Verified that we're still pending a installer template update.

Merging in the latest changes in `main` (or rebasing) would help us reason about and validate the patch.

Now that `main` is tracking v1.9 development work, it's a good time to get this PR in final shape. @elfenlaid are you able to take this further?

elfenlaid commented:
> We should also remove {:jason, "~> 1.2"} from new apps.
> Verified that we're still pending a installer template update.

I'm not sure if that is possible to do with the current installer's requirements. Currently the installer requires 1.17, but JSON landed in 1.18 if I'm not wrong 🤔

josevalim commented:
We can bump the installer requirement too!

elfenlaid commented:
Oh nice, I think I can handle it then 😅

elfenlaid commented:
Oof, there's another problem, aside from Elixir 1.18+ we should also require Erlang/OTP 27+ for that to work -_-

Currently the installation instructions ask for Erlang/OTP 25+

josevalim commented:
26 and earlier are no longer officially supported, so we should bump that too!

elfenlaid commented on guides/introduction/installation.md:
```diff
@@ -2,33 +2,33 @@
 
 In order to build a Phoenix application, we will need a few dependencies installed in our Operating System:
 
-  * the Erlang VM and the Elixir programming language
```
Sorry, this was an automatic fix from the markdown autoformatter. I can revert it, but it seems like acceptable change, unless those asterisks where intentional.

elfenlaid commented on installer/lib/mix/tasks/phx.new.ex:
```diff
@@ -565,66 +564,61 @@ defmodule Mix.Tasks.Phx.New do
   end
 
-  # we need to parse JSON, so we only check for new versions on Elixir 1.18+
-  if Version.match?(System.version(), "~> 1.18") do
```
Maybe this a too bold of a change, but it kinda makes sense if we insist on having Elixir 1.18+ / OTP 27 for the installer.

rhcarvalho commented on config/config.exs:
```diff
@@ -5,7 +5,8 @@ config :logger, :console,
 config :phoenix,
-  json_library: Jason,
+  # TODO: Remove the `json_library` check once `JSON` becomes the standard `Phoenix.json_library/1`
+  json_library: (if Code.ensure_loaded?(JSON), do: JSON, else: Jason),
```
Now that we decided to bump Elixir/OTP version requirements, maybe we don't need the comment and conditional?

elfenlaid replied:
This would be a breaking change as far as I can say. The installer creates a new app, while this is used in existing apps.

josevalim replied:
config/config.exs is not used by dependencies. So we can likely default to JSON, no problems.

SteffenDE replied:
It is used by tests though and we still test with Elixir 1.15

rhcarvalho replied:
In that case we can just update the comment. We only need the conditional while Phoenix itself is tested against older Elixir/OTP, unrelated to any particular installer defaults.

elfenlaid commented:
Okay, I've bumped the installer version to 1.18 and I think the PR is ready for yet another review pass.

I've tried to scope changes in commits and avoid collateral damage, but I understand that it's not that easy to follow given how the PR evolved. I can extract the installer changes to a separate PR if needed, just let me know.

josevalim commented on guides/controllers.md:
```diff
@@ -98,7 +98,7 @@ end
 Now [`/hello/Frank`] in your browser should display `From messenger Frank` as plain text without any HTML.
 
-A step beyond this is rendering pure JSON with the [`json/2`] function. We need to pass it something that the [Jason library](`Jason`) can decode into JSON, such as a map. (Jason is one of Phoenix's dependencies.)
+A step beyond this is rendering pure JSON with the [`json/2`] function. We need to pass it something that the [JSON library](https://elixir.hexdocs.pm/JSON.html) can decode into JSON, such as a map.
```
```suggestion
A step beyond this is rendering pure JSON with the [`json/2`] function. We need to pass it something that the [JSON](https://elixir.hexdocs.pm/JSON.html) module can encode, such as a map.
```

josevalim commented on guides/json_and_apis.md:
````diff
@@ -175,7 +175,7 @@ defmodule HelloWeb.UrlJSON do
 ```
 
-This view is very simple. The `index` function receives all URLs, and converts them into a list of maps. Those maps are placed inside the data key at the root, exactly as we saw when interfacing with our application from `cURL`. In other words, our JSON view converts our complex data into simple Elixir data-structures. Once our view layer returns, Phoenix uses the `Jason` library to encode JSON a ...
+This view is very simple. The `index` function receives all URLs, and converts them into a list of maps. Those maps are placed inside the data key at the root, exactly as we saw when interfacing with our application from `cURL`. In other words, our JSON view converts our complex data into simple Elixir data-structures. Once our view layer returns, Phoenix uses the built-in [JSON library](https:// ...
````
```suggestion
This view is very simple. The `index` function receives all URLs, and converts them into a list of maps. Those maps are placed inside the data key at the root, exactly as we saw when interfacing with our application from `cURL`. In other words, our JSON view converts our complex data into simple Elixir data-structures. Once our view layer returns, Phoenix uses the built-in [JSON module](https://elixir.hexdocs.pm/JSON.html) to encode JSON and send the response to the client.
```

josevalim reviewed (commented):
This looks good to me. I have dropped some comments. I believe we also need to configure Ecto to use JSON.

josevalim commented:
>  I believe we also need to configure Ecto to use JSON.

To be more precise, we nee to configure the underlying adapter (postgrex, myxql, tds, etc). We can do it in the same place we call `config :phoenix, :json_library, JSON`. We just swap the app name.

elfenlaid commented on installer/templates/phx_single/config/config.exs.eex:
```diff
@@ -76,6 +76,28 @@ config :logger, :default_formatter,
 # Use built-in `JSON` for JSON parsing in Phoenix
 config :phoenix, :json_library, JSON
 
+<%= if @ecto do %>
```
Not the most elegant solution for sure 🙈

josevalim replied:
The best option is to define a variable in the generators themselves, something like @ecto_adapter_app or similar, anid then in here you do something like:

```
config :phoenix, :json_library, JSON<%= if @ecto_adapter_app do %>
config :<%= @ecto_adapter_app %>, :json_library, JSON<% end %>
```

elfenlaid replied:
Oh, what's more, there's `adapter_app`, I haven't noticed it 😅

elfenlaid commented:
> This looks good to me. I have dropped some comments. I believe we also need to configure Ecto to use JSON.

Thanks, I've merged them and added the adapters configuration.

josevalim commented on installer/templates/phx_single/config/config.exs.eex:
```diff
@@ -73,9 +73,14 @@ config :logger, :default_formatter,
+# Use built-in `JSON` for JSON coding in Ecto
+config :<%= @adapter_app %>, :json_library, JSON
+
+<% end %>
```
So we don't have extra newlines:

```suggestion
config :phoenix, :json_library, JSON<%= if @ecto do %>
config :<%= @adapter_app %>, :json_library, JSON<% end %>
```

josevalim commented on installer/templates/phx_umbrella/config/extra_config.exs.eex:
```diff
@@ -5,8 +5,8 @@ config :logger, :default_formatter,
-# Use Jason for JSON parsing in Phoenix
-config :phoenix, :json_library, Jason
+# Use built-in `JSON` for JSON parsing in Phoenix
+config :phoenix, :json_library, JSON
```
We also need the ecto change here!

elfenlaid replied:
Good catch and sorry for stealing so much of your time 🙇

josevalim commented on installer/templates/phx_single/config/config.exs.eex:
```diff
@@ -73,9 +73,9 @@ config :logger, :default_formatter,
-
+# Use built-in `JSON` for JSON parsing in Phoenix
+config :phoenix, :json_library, JSON<%= if @ecto do %>
+config :<%= @adapter_app %>, :json_library, JSON<% end %>
```
```suggestion
# Use built-in `JSON` module
config :phoenix, :json_library, JSON<%= if @ecto do %>
config :<%= @adapter_app %>, :json_library, JSON<% end %>

```

SteffenDE commented on config/config.exs:
```diff
@@ -5,7 +5,8 @@ config :logger, :console,
 config :phoenix,
-  json_library: Jason,
+  # TODO: Remove the `json_library` check once `JSON` becomes the standard `Phoenix.json_library/1`
+  json_library: (if Code.ensure_loaded?(JSON), do: JSON, else: Jason),
```
```suggestion
  # TODO: Remove once we don't test with Elixir < 1.18 any more
  json_library: (if Code.ensure_loaded?(JSON), do: JSON, else: Jason),
```

SteffenDE commented:
Thank you! 🙌🏻
